### PR TITLE
tka: implement consensus & state computation internals

### PR DESCRIPTION
**Pull request by twitchyliquid64**
This PR implements the nitty-gritty consensus internals. `fastForward` and `computeStateAt` are the 'core' algorithms that drive the whole thing.

A lot of the delta are tests or the new test helper `chaintest_test.go`.

### CODE REVIEW

File: tka/tka.go [hunk 1]
```diff
@@ -4,3 +4,331 @@
...
+func pickNextAUM(state State, candidates []AUM) AUM {
+	switch len(candidates) {
+	case 0:
+		panic("pickNextAUM called with empty candidate set")
+	case 1:
+		return candidates[0]
+	}
+
+	// Oooof, we have some forks in the chain. We need to pick which
+	// one to use by applying the Fork Resolution Algorithm :sparkles:
```
Comment (crawshaw) [inline tka/tka.go:99]: it's utf8, you can paste ✨ into the comment :)

File: tka/tka.go [hunk 2]
```diff
@@ -4,3 +4,331 @@
...
+	//
+	// The rules are this:
+	// 1. The child with the highest signature weight is chosen.
```
Comment (crawshaw) [inline tka/tka.go:102]: You could inline these comments below

File: tka/tka.go [hunk 3]
```diff
@@ -4,3 +4,331 @@
...
+	// Not strictly necessary, but simplifies checks in tests.
+	sort.Slice(candidates, func(i, j int) bool {
+		ih, jh := candidates[i].Oldest.Hash(), candidates[j].Oldest.Hash()
+		return bytes.Compare(ih[:], jh[:]) < 0
+	})
+
+	// candidates.Oldest needs to be computed by working backwards from
+	// head as far as we can.
+	iterAgain := true // if theres still work to be done.
+	for i := 0; i < maxIter && iterAgain; i++ {
```
Comment (crawshaw) [inline tka/tka.go:56]: what happens if you hit maxIter? no need to error out? (it's passed into computeActiveChain and could use some docs)

File: tka/tka.go [hunk 4]
```diff
@@ -4,3 +4,331 @@
...
+// ancestor of the valid chain.
+//
+// If all the chains end up having the same ancestor, then thats the
+// only possible ancestor, ezpz. However if there are multiple distinct
+// ancestors, that means there are distinct chains, and we need some
+// hint to choose what to use. For that, we rely on the chainsThroughActive
+// bit, which signals to us that that ancestor was part of the
+// chain in a previous run.
+func computeActiveAncestor(storage Chonk, chains []chain) (AUMHash, error) {
+	// Dedupe possible ancestors, tracking if they were part of
```
Comment (crawshaw) [inline tka/tka.go:262]: why would there be a duplicate? it feels like there's something to `computeChainCandidates` I don't understand yet

**Review by crawshaw — APPROVED** [review]
I'm not convinced I really understand `computeChainCandidates`, might need to talk it through sometime, but after that everything seems straightforward.

File: tka/tka.go (hunk 1 above)
Comment (twitchyliquid64) [reply tka/tka.go:99]: VIBES ✨

File: tka/tka.go (hunk 3 above)
Comment (twitchyliquid64) [reply tka/tka.go:56]: Welp missed that, added!

**Comment by twitchyliquid64** [comment]
If everything goes well and storage is exactly how we expect it, then there will never be more than one distinct chain (ie: all chains have a common ancestor). `computeChainCandidates()` and `computeActiveAncestor()` is all about being resilient to if that isnt the case for whatever reason.

I have half a mind to pull that out: just compute the active chain from the last known ancestor (as per the TODO).

Does that make more sense?

**Comment by crawshaw** [comment]
Don’t you still need the candidates concept for the case where you get one aum then the server rolls back and you get a different one sent to you?

**Comment by crawshaw** [comment]
I see. Probably worth keeping for that case then.

**Pull request merged**